src/main/java/WayofTime/bloodmagic/compress/AdvancedCompressionHandler.java

```diff
@@ -4,12 +4,82 @@
 import net.minecraft.inventory.Container;
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.ItemStack;
+import net.minecraft.item.crafting.CraftingManager;
+import net.minecraft.item.crafting.IRecipe;
 import net.minecraft.world.World;
 
+
 public class AdvancedCompressionHandler extends CompressionHandler {
+    private static ItemStack reversibleCheck;
+
+    public static boolean isResultStackReversible(ItemStack stack, int gridSize, World world) {
+
+        if (stack.isEmpty()) {
+            return false;
+        }
+        InventoryCrafting inventory = new InventoryCrafting(new Container() {
+            public boolean canInteractWith(EntityPlayer player) {
+                return false;
+            }
+        }, 2, 2);
+
+        inventory.setInventorySlotContents(0, stack);
+        ItemStack returnStack = getNNRecipeOutput(inventory, world);
+
+        return !returnStack.isEmpty() && CompressionRegistry.areItemStacksEqual(reversibleCheck, returnStack);
+    }
+
+    public static ItemStack getRecipe(ItemStack stack, World world, int gridSize) {
+        StorageBlockCraftingManager craftingManagerSB = StorageBlockCraftingManager.getInstance();
+        InventoryCrafting inventory = new InventoryCrafting(new Container() {
+            public boolean canInteractWith(EntityPlayer player) {
+                return false;
+            }
+        }, gridSize, gridSize);
+        for (int i = 0; i < inventory.getSizeInventory(); i++) {
+            inventory.setInventorySlotContents(i, stack);
+        }
+        ItemStack notEmptyRecipe = craftingManagerSB.findMatchingRecipe(inventory, world);
+        if (!notEmptyRecipe.isEmpty()) {
+            return notEmptyRecipe;
+        }
+        ItemStack result = getNNRecipeOutput(inventory, world);
+
+        if (isResultStackReversible(result, gridSize, world)) {
+            craftingManagerSB.addRecipe(CraftingManager.findMatchingRecipe(inventory, world));
+            return result;
+        }
+        return ItemStack.EMPTY;
+    }
+
+    public static ItemStack getNNRecipeOutput(InventoryCrafting inventory, World world) {
+        IRecipe checkForNull = CraftingManager.findMatchingRecipe(inventory, world);
+        if (checkForNull != null) {
+            return checkForNull.getRecipeOutput();
+        }
+        return ItemStack.EMPTY;
+    }
+
+    public static boolean has22Recipe(ItemStack stack, World world) {
```

Comment:
Remove this. Anybody using this is being inefficient since they can just call `get22Recipe(...)` and check if it's empty themselves.